From a3ffd4c8c904e4225d8543054667d2b5ebe34052 Mon Sep 17 00:00:00 2001
From: VinceChang6637 <vince_chang@aspeedtech.com>
Date: Wed, 29 Nov 2023 11:19:03 +0800
Subject: [PATCH 4/4] obmc-ikvm: support ast2750 dual nodes.

Expect result:
/usr/bin/obmc-ikvm -v /dev/video0 -k /dev/hidg0 -p /dev/hidg1
/usr/bin/obmc-ikvm -v /dev/video1 -k /dev/hidg2 -p /dev/hidg3

/dev/video0 listen port 5900
/dev/video1 listen port 5901
/dev/hidg0 use "12011000.usb-vhub" UDC
/dev/hidg2 use "12021000.usb-vhub" UDC

Upstream-Status: Pending
Signed-off-by: VinceChang6637 <vince_chang@aspeedtech.com>
---
 ikvm_input.cpp  | 12 ++++++++++++
 ikvm_input.hpp  | 14 +++++++++++---
 ikvm_server.cpp |  6 ++++++
 3 files changed, 29 insertions(+), 3 deletions(-)

Index: git/ikvm_input.cpp
===================================================================
--- git.orig/ikvm_input.cpp
+++ git/ikvm_input.cpp
@@ -28,6 +28,18 @@ Input::Input(const std::string& kbdPath,
     pointerFd(-1), keyboardReport{0}, pointerReport{0}, keyboardPath(kbdPath),
     pointerPath(ptrPath), udcName(udc)
 {
+    // Support AST2750 dual nodes
+    if (kbdPath == "/dev/hidg2")
+    {
+        hidUdcPath = OBMC_HID_PATH_1;
+        usbVirtualHubPath = USB_VHUB_PATH_1;
+    }
+    else
+    {
+        hidUdcPath = OBMC_HID_PATH;
+        usbVirtualHubPath = USB_VHUB_PATH;
+    }
+
     hidUdcStream.exceptions(std::ofstream::failbit | std::ofstream::badbit);
     hidUdcStream.open(hidUdcPath, std::ios::out | std::ios::app);
 }
Index: git/ikvm_input.hpp
===================================================================
--- git.orig/ikvm_input.hpp
+++ git/ikvm_input.hpp
@@ -79,11 +79,15 @@ class Input
         0x40  // right alt
     };
     /* @brief Path to the HID gadget UDC */
-    static constexpr const char* hidUdcPath =
+    static constexpr const char* OBMC_HID_PATH =
         "/sys/kernel/config/usb_gadget/obmc_hid/UDC";
+    static constexpr const char* OBMC_HID_PATH_1 =
+        "/sys/kernel/config/usb_gadget/obmc_hid1/UDC";
     /* @brief Path to the USB virtual hub */
-    static constexpr const char* usbVirtualHubPath =
-        "/sys/bus/platform/devices/1e6a0000.usb-vhub";
+    static constexpr const char* USB_VHUB_PATH =
+        "/sys/bus/platform/devices/12011000.usb-vhub";
+    static constexpr const char* USB_VHUB_PATH_1 =
+        "/sys/bus/platform/devices/12021000.usb-vhub";
     /* @brief Retry limit for writing an HID report */
     static constexpr int HID_REPORT_RETRY_MAX = 5;
     /*
@@ -116,6 +120,10 @@ class Input
     std::string pointerPath;
     /* @brief Name of UDC */
     std::string udcName;
+    /* @brief Path to the HID gadget UDC */
+    std::string hidUdcPath;
+    /* @brief Path to the USB virtual hub */
+    std::string usbVirtualHubPath;
     /*
      * @brief Mapping of RFB key code to report data index to keep track
      *        of which keys are down
Index: git/ikvm_server.cpp
===================================================================
--- git.orig/ikvm_server.cpp
+++ git/ikvm_server.cpp
@@ -49,6 +49,12 @@ Server::Server(const Args& args, Input&
     server->cursor->xhot = 1;
     server->cursor->yhot = 1;
 
+    // Support AST2750 dual nodes
+    if (args.getVideoPath() == "/dev/video1")
+    {
+        server->port = 5901;
+    }
+
     rfbStringToAddr(&ip[0], &server->listenInterface);
 
     rfbInitServer(server);
